clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1607,6 +1607,7 @@ private Fetch<K, V> pollForFetches(Timer timer) {
             wakeupTrigger.clearTask();
         }
 
+        metadata.maybeThrowAnyException();
```

Comment:
uhm I would say that having this here, in the app thread, is not right, since the metadata updates happen in the background thread, so we could easily get a race condition I would expect. The metadata object is updated in the [NetworkClient](https://github.com/apache/kafka/blob/bb3ff0f84a0a4c6d213fbe50eb2c3d3a8cdecdf0/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L1193), when a response to a metadata request is received. So only at that point in the background thread is that may accurately know that there is an invalid topic. I imagine we should then propagate an `ErrorEvent` to the app thread, that would be raised in the poll loop [here](https://github.com/apache/kafka/blob/bb3ff0f84a0a4c6d213fbe50eb2c3d3a8cdecdf0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L213))

Seen from a conceptual POV, we have something that could wrong in the background (invalid topic received in metadata response), and we want to throw an error in the foreground, so the pattern would be via events. Similar probably to what is done to raise errors that may happen while resetting positions in the background thread, that are propagated to the app thread [here](https://github.com/apache/kafka/blob/bb3ff0f84a0a4c6d213fbe50eb2c3d3a8cdecdf0/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L202). Difference in this new case is that the failure is detected deep down the network client (not in the request managers layer), so if we follow this approach, we would need some wiring there. Just thinking out loud, this is definitely a tricky missing bit

Reply:
Thank you for review. @lianetm 
I learned a lot from it.

It seems that we can solve the wiring issue using MetadataUpdater in this case.
What do you think?

I need more time for this issue and I don't think we have a lot of time left for this issue because it's a blocker.
so it's fine if you hand it off to someone else.

Reply:
@lianetm Based on the feedback in the review, I've developed a proposal for using the MetadataUpdater.
However, test code still needs to be written. I've developed this as a draft for now